protocol/v2/types/batch_verifier.go

```diff
@@ -0,0 +1,475 @@
+		b.timer.Stop()
+		previouslyStarted := b.started
+		b.started = time.Time{}
+		pending := b.pending.Values()
+		b.pending.Reset()
+		b.mu.Unlock()
+
+		b.debug.Lock()
+		b.debug.batchPendingDurations.AddTime(time.Since(previouslyStarted))
+		b.debug.Unlock()
+
+		b.batches <- batch{
+			start:    previouslyStarted,
+			requests: pending,
+		}
+	} else {
+		// Batch has grown: adjust the timer.
+		b.timer.Stop()
+		t := b.adaptiveTimeout(b.pending.Len())
+		b.timer.Reset(t)
+		b.started = time.Now()
+		b.mu.Unlock()
+	}
+
+	valid := <-result
+	if !valid {
+		// If the batch failed, fall back to individual verification.
+		return b.verifySingle(sr)
+	}
+	return valid
+}
+
+// adaptiveTimeout calculates the timeout based on the proportion of pending requests.
+func (b *BatchVerifier) adaptiveTimeout(pending int) time.Duration {
+	if b.started.IsZero() {
+		b.started = time.Now()
+	}
+	workload := int(b.busyWorkers.Load()) + len(b.batches) + int(float64(pending)/float64(b.batchSize))
+	if workload > b.concurrency {
+		workload = b.concurrency
+	}
+	// busyness := float64(workload) / float64(b.concurrency) / 2
+	busyness := float64(workload+1) / float64(b.concurrency) * 2
+	if busyness > 1 {
+		busyness = 1
+	}
+
+	timeLeft := b.timeout - time.Since(b.started)
+	if timeLeft <= 0 {
+		return 0
+	}
+	// log.Printf("workload: %d, busyness: %f, timeLeft: %s, adjusted: %s", workload, busyness, timeLeft, time.Duration(busyness*float64(timeLeft)))
+	return time.Duration(busyness * float64(timeLeft))
+}
+
+// Start launches the worker goroutines.
+func (b *BatchVerifier) Start() {
+	go func() {
+		for {
+			time.Sleep(12 * time.Second)
+			stats := b.Stats()
+			formatGauge := func(g Gauge[time.Duration]) string {
+				ms := func(d time.Duration) string {
+					return fmt.Sprintf("%.1fms", d.Seconds()*1000)
+				}
+				return fmt.Sprintf("min: %s,\nmax: %s,\navg: %s,\np99: %s,\np95: %s,\np50: %s", ms(g.Min), ms(g.Max), ms(g.Avg), ms(g.P99), ms(g.P95), ms(g.P50))
+			}
+			zap.L().Debug("BatchVerifier stats",
+				zap.Int("total_requests", stats.TotalRequests),
+				zap.Int("duplicate_requests", stats.DuplicateRequests),
+				zap.Int("total_batches", stats.TotalBatches),
+				zap.Float64("average_batch_size", stats.AverageBatchSize),
+				zap.Int("pending_requests", stats.PendingRequests),
+				zap.Int("pending_batches", stats.PendingBatches),
+				zap.Int("busy_workers", stats.BusyWorkers),
+				zap.Any("recent_batch_sizes", stats.RecentBatchSizes),
+				zap.Int("failed_batches", stats.FailedBatches),
+				zap.Int("failed_requests", stats.FailedRequests),
+				zap.String("request_total_durations", formatGauge(stats.RequestTotalDurations)),
+				zap.String("request_pending_durations", formatGauge(stats.RequestPendingDurations)),
+				zap.String("batch_durations", formatGauge(stats.BatchDurations)),
+				zap.String("batch_pending_durations", formatGauge(stats.BatchPendingDurations)),
+			)
+		}
+	}()
+
```

Comment:
I would move this code to a separate func, maybe add a todo about making this optional

Reply:
Generally, I agree, but I think we don't need to touch the batch verifier anymore because we'll likely remove it. I'd rather delete the file but @moshe-blox wanted to keep it